### ISSUE TITLE
CompletionStageTimeout uses CompletableFuture.cancel(true) incorrectly

### ISSUE THREAD

**Issue by Ladicek** [smallrye/smallrye-fault-tolerance#213]
The `CompletionStageTimeout` strategy attempts to interrupt an execution thread when timeout occurs using `CompletableFuture.cancel(true)`. However, `CompletableFuture` never interrupts any thread (there doesn't even have to be one).

This needs to be refactored. Perhaps we can use the out-of-band communication capability of `InvocationContext` to pass the "current thread" object from the invocation back to the `CompletionStageTimeout`.

**Comment by Ladicek** [comment]
I've realized one thing: cancelling a `CompletionStage`-based async computation isn't 100% possible.

What is possible is this simple situation:

```java
@Asynchronous
public CompletionStage<String> foobar() {
    ... some lenghty _synchronous_ computation ...
    return completedFuture(...);
}
```

Here, the `@Asynchronous` annotation is used to offload a synchronous computation to a different thread. We could possibly interrupt that thread, when timeout occurs.

However, if someone calls yet another `CompletionStage`-based asynchronous computation, we're screwed. Something like this:

```java
@Asynchronous
public CompletionStage<String> foobar() {
    return someAsyncWork().thenCompose(x -> someOtherAsyncWork(x));
}
```

Here, the method will basically immediately return, and the async computations will most likely run on yet another threads. In this case, cancellation is impossible, because `CompletionStage` just doesn't have that feature.

**Comment by Ladicek** [comment]
Fixed in #309.
